configuration/postgres/postgres.go

```diff
@@ -0,0 +1,415 @@
+		}
+		return nil, err
+	}
+	items := make(map[string]*configuration.Item)
+	for i := 0; rows.Next(); i++ {
+		var item configuration.Item
+		var key string
+		var metadata []byte
+		v := make(map[string]string)
+
+		if err := rows.Scan(&key, &item.Value, &item.Version, &metadata); err != nil {
+			return nil, err
+		}
+		if err := json.Unmarshal(metadata, &v); err != nil {
+			return nil, err
+		}
+		item.Metadata = v
+		if item.Value != "" {
+			items[key] = &item
+		}
+	}
+	return &configuration.GetResponse{
+		Items: items,
+	}, nil
+}
+
+func (p *ConfigurationStore) Subscribe(ctx context.Context, req *configuration.SubscribeRequest, handler configuration.UpdateHandler) (string, error) {
+	var triggers []string
+	for k, v := range req.Metadata {
+		if res := strings.EqualFold("trigger", k); res {
+			triggers = append(triggers, v)
+		}
+	}
+
+	if len(triggers) == 0 {
+		return "", fmt.Errorf("cannot subscribe to empty trigger")
+	}
+
+	var subscribeID string
+	p.configLock.Lock()
+	for _, trigger := range triggers {
+		notificationChannel := "listen " + trigger
+		if sub, isActive := p.isSubscriptionActive(req); isActive {
+			if oldStopChan, ok := p.subscribeStopChanMap[sub]; ok {
+				close(oldStopChan.(chan struct{}))
+			}
+		}
+		stop := make(chan struct{})
+		subscribeID = uuid.New().String()
+		p.subscribeStopChanMap[subscribeID] = stop
+		p.ActiveSubscriptions[trigger] = &subscription{
+			uuid:    subscribeID,
+			trigger: trigger,
+			keys:    req.Keys,
+		}
+		go p.doSubscribe(ctx, req, handler, notificationChannel, trigger, subscribeID, stop)
+	}
+	p.configLock.Unlock()
+	return subscribeID, nil
+}
+
+func (p *ConfigurationStore) Unsubscribe(ctx context.Context, req *configuration.UnsubscribeRequest) error {
+	p.configLock.Lock()
+	defer p.configLock.Unlock()
+	if oldStopChan, ok := p.subscribeStopChanMap[req.ID]; ok {
+		delete(p.subscribeStopChanMap, req.ID)
+		close(oldStopChan.(chan struct{}))
+		for k, v := range p.ActiveSubscriptions {
+			if v.uuid == req.ID {
+				channel := "unlisten " + v.trigger
+				conn, err := p.client.Acquire(ctx)
+				if err != nil {
+					return fmt.Errorf("Error acquiring connection: %v", err)
+				}
+				defer conn.Release()
+				_, err = conn.Exec(ctx, channel)
+				if err != nil {
+					return fmt.Errorf("Error unlistening to channel: %v", err)
+				}
+				delete(p.ActiveSubscriptions, k)
+			}
+		}
+		return nil
+	}
+	return fmt.Errorf("unable to find subscription with ID : %v", req.ID)
+}
+
+func (p *ConfigurationStore) doSubscribe(ctx context.Context, req *configuration.SubscribeRequest, handler configuration.UpdateHandler, channel string, trigger string, subscription string, stop chan struct{}) {
+	conn, err := p.client.Acquire(ctx)
+	if err != nil {
+		p.logger.Errorf("Error acquiring connection:", err)
+		return
+	}
+	defer conn.Release()
+	_, err = conn.Exec(ctx, channel)
+	if err != nil {
+		p.logger.Errorf("Error listening to channel:", err)
+		return
+	}
+
```

Comment:
Cleaner to split ! Across both conditions and inverse the or to and?

Reply:
Done

Reply:
https://github.com/dapr/components-contrib/blob/90ba8a441d6ccc8987acf115ca4adf1f49fa9fba/configuration/postgres/postgres.go#L225-L231